src/synapse/api.py

```diff
@@ -151,18 +162,31 @@ def _get_nonce() -> str:
 
     Raises:
         NetworkError: if there was an error fetching the nonce.
+        GetNonceError: if there was an error while getting nonce.
     """
     try:
         res = requests.get(REGISTER_URL, timeout=5)
         res.raise_for_status()
-        return res.json()["nonce"]
-    except (
-        requests.exceptions.ConnectionError,
-        requests.exceptions.Timeout,
-        requests.exceptions.HTTPError,
-    ) as exc:
-        logger.exception("Failed to request %s : %r", REGISTER_URL, exc)
-        raise NetworkError(f"Failed to request {REGISTER_URL}.") from exc
+        res_json = res.json()
+        if not isinstance(res_json, dict):
+            # Exception not in docstring because is captured.
+            raise GetNonceError(f"Response has unexpected encode: {res_json}")  # noqa: DCO053
+        nonce = res_json.get("nonce", None)
+        if nonce is None:
+            # Exception not in docstring because is captured.
+            raise NonceNotFoundError(  # noqa: DCO053
+                f"There is no nonce in JSON output: {res_json}"
+            )
+        return nonce
+    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as exc:
+        logger.exception("Failed to connect to %s: %r", REGISTER_URL, exc)
+        raise NetworkError(f"Failed to connect to {REGISTER_URL}.") from exc
+    except requests.exceptions.HTTPError as exc:
+        logger.exception("HTTP error from %s: %r", REGISTER_URL, exc)
+        raise NetworkError(f"HTTP error from {REGISTER_URL}.") from exc
+    except (GetNonceError, requests.exceptions.JSONDecodeError) as exc:
+        logger.exception("Failed to get nonce: %r", exc)
+        raise GetNonceError(str(exc)) from exc
```

Comment:
I'll try and have a go, see what you think:

```python
   try:
        res = requests.get(REGISTER_URL, timeout=5)
        res.raise_for_status()
        res_json = res.json()
    except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as exc:
        logger.exception("Failed to connect to %s: %r", REGISTER_URL, exc)
        raise NetworkError(f"Failed to connect to {REGISTER_URL}.") from exc
    except requests.exceptions.HTTPError as exc:
        logger.exception("HTTP error from %s: %r", REGISTER_URL, exc)
        raise NetworkError(f"HTTP error from {REGISTER_URL}.") from exc
    # TODO: I think there are also possible exceptions for res.json(), I'm not sure what they are, e.g., if the response was 201 I'm not sure what the behaviour of res.json() is

    if not isinstance(res_json, dict):
        raise GetNonceError(f"Response has unexpected encode: {res_json}")
    nonce = res_json.get("nonce", None)
    if nonce is None:
        raise NonceNotFoundError(
            f"There is no nonce in JSON output: {res_json}"
        )
    return nonce
```

Reply:
You made me realize that it was really confusing before. Now I changed everything based on your idea. Please take a look.
I'm just concerned that I couldn't add Exceptions in docstring that I know will be raised but the lint says it should not be documented.

Reply:
Since the code is updated, I'll close this review but feel free to reopen it.